### DISCUSSION TITLE
zfs destroy not working

### DISCUSSION THREAD

**Discussion by spflanze** [openzfs/zfs#14525] [Q&A] [answered]
The zfs destroy command is not working for me. The command sequence that illustrates this:

```
stephen@stephen:~$ zfs list -t snapshot -o name | grep ^bpool/BOOT/ubuntu_7pdn8o@autozsys_
bpool/BOOT/ubuntu_7pdn8o@autozsys_exlhhf
bpool/BOOT/ubuntu_7pdn8o@autozsys_zrsqqy
bpool/BOOT/ubuntu_7pdn8o@autozsys_1dxbpl
bpool/BOOT/ubuntu_7pdn8o@autozsys_b6ci1z
bpool/BOOT/ubuntu_7pdn8o@autozsys_3jghsv

stephen@stephen:~$ sudo zfs destroy -rd bpool/BOOT/ubuntu_7pdn8o@autozsys_exlhhf

stephen@stephen:~$ zfs list -t snapshot -o name | grep ^bpool/BOOT/ubuntu_7pdn8o@autozsys_
bpool/BOOT/ubuntu_7pdn8o@autozsys_exlhhf
bpool/BOOT/ubuntu_7pdn8o@autozsys_zrsqqy
bpool/BOOT/ubuntu_7pdn8o@autozsys_1dxbpl
bpool/BOOT/ubuntu_7pdn8o@autozsys_b6ci1z
bpool/BOOT/ubuntu_7pdn8o@autozsys_3jghsv
stephen@stephen:~$ 
```
As can be seen the snapshot "bpool/BOOT/ubuntu_7pdn8o@autozsys_exlhhf" did not get destroyed. It persists even after reboot. What went wrong here?

There are now too many snapshots in bpool. This is preventing system updates. My inability to delete the oldest snapshots leaves me unable to solve this problem, and get my system updated.

**Comment by bghira** [comment, marked as answer]
`-d` indicates deferred destroy, try without that.

**Reply by spflanze** [reply]
No. According to https://openzfs.github.io/openzfs-docs/man/8/zfs-destroy.8.html#d :

```
-d
    Destroy immediately. If a snapshot cannot be destroyed now, mark it for deferred destruction.
```

**Reply by bghira** [reply]
> ... , mark it for deferred destruction

sounds like it does what i described. can you please try to destroy without `-d`?

**Reply by spflanze** [reply]
It worked to remove the -d. Thanks :). The command I used is:
`sudo zfs destroy -rR bpool/BOOT/ubuntu_7pdn8o@autozsys_exlhhf`
The -R is because there was an error message that there were dependent clones.

**Reply by rincebrain** [reply]
That sounds like the actual problem might be that -d masks the -R needed error.
